### PR TITLE
add BaseModel; store type in serialization

### PR DESCRIPTION
**Pull request by njzjz**

### CODE REVIEW

File: deepmd/dpmodel/model/base_model.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+from deepmd.utils.plugin import (
+    Plugin,
+)
+
+
+class BaseBaseModel(ABC):
+    """Base class for final exported model that will be directly used for inference.
+
+    The class defines some abstractmethods that will be directly called by the
+    inference interface. If the final model class inherbits some of those methods
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/model/base_model.py:25]:
```suggestion
    inference interface. If the final model class inherits some of those methods
```

File: deepmd/pt/model/model/model.py [hunk 2]
```diff
@@ -36,3 +149,22 @@ def compute_or_load_stat(
...
+    @abstractmethod
+    def get_nnei(self) -> int:
+        """Returns the total number of selected neighboring atoms in the cut-off radius."""
+        # for C++ interface
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_nsel(self) -> int:
+        """Returns the total number of selected neighboring atoms in the cut-off radius."""
+        raise NotImplementedError
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/model/model.py:170]:
```suggestion
        pass
```

File: deepmd/pt/model/model/model.py [hunk 3]
```diff
@@ -36,3 +149,22 @@ def compute_or_load_stat(
...
+    @abstractmethod
+    def get_model_def_script(self) -> str:
+        """Get the model definition script."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_nnei(self) -> int:
+        """Returns the total number of selected neighboring atoms in the cut-off radius."""
+        # for C++ interface
+        raise NotImplementedError
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/model/model.py:165]:
```suggestion
              pass
```

File: deepmd/pt/model/model/model.py [hunk 4]
```diff
@@ -36,3 +149,22 @@ def compute_or_load_stat(
...
         """
         raise NotImplementedError
+
+    model_def_script: str
+
+    # currently, only pt needs the following methods
+    @abstractmethod
+    def get_model_def_script(self) -> str:
+        """Get the model definition script."""
+        raise NotImplementedError
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/model/model.py:159]:
```suggestion
        pass
```

**Review by wanghan-iapcm — COMMENTED** [review]
- I do not understand why the `serialization` and `deserialization` cannot be put int `BaseBaseModel`. 
- would the following inheritance work?
```
#       torch.nn.Module        BaseAtomicModel        BaseBaseModel
#             |                       |                    |
#             -------------------------                    |
#                         |                                |
#                    DPAtomicModel                         |
#                         |                                |
#                make_model(DPAtomicModel)                 |
#                         |                                |
#                         ----------------------------------
#                                           |
#                                         DPModel
```
- in our concept a model is an atomic model, therefore, the base model seems to be a facilitate that provides the plugin, other methods like the `get_rcut`, `get_dim_fparam`... seems to be inherited from atomic model rather than model .

**Comment by njzjz** [comment]
> * I do not understand why the `serialization` and `deserialization` cannot be put int `BaseBaseModel`.

Done.

> * would the following inheritance work?
> 
> ```
> #       torch.nn.Module        BaseAtomicModel        BaseBaseModel
> #             |                       |                    |
> #             -------------------------                    |
> #                         |                                |
> #                    DPAtomicModel                         |
> #                         |                                |
> #                make_model(DPAtomicModel)                 |
> #                         |                                |
> #                         ----------------------------------
> #                                           |
> #                                         DPModel
> ```
> 

We cannot store the plugins of dpmodel and pt in the same class. There must be a class to initialize the plugin separately.

> * in our concept a model is an atomic model, therefore, the base model seems to be a facilitate that provides the plugin, other methods like the `get_rcut`, `get_dim_fparam`... seems to be inherited from atomic model rather than model .

It looks to me `SpinModel` in #3301 is not inherited from an atomic model.

File: deepmd/pt/model/model/model.py [hunk 5]
```diff
@@ -1,16 +1,111 @@
...
+    Plugin,
+)
 
 
-class BaseModel(torch.nn.Module):
+# trick: torch.nn.Module should not be inherbited here, otherwise,
+# the abstract method will override the method from the atomic model
+# as Python resolves method lookups using the C3 linearisation.
+# See https://stackoverflow.com/a/47117600/9567349
+# Take an example, this is sitatuion for only inherbiting BaseBaseModel:
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pt/model/model/model.py:29]:
```suggestion
# Take an example, this is the situation for only inheriting BaseBaseModel:
```

File: deepmd/pt/model/model/model.py [hunk 6]
```diff
@@ -1,16 +1,111 @@
...
+#                    DPAtomicModel                      BaseModel
+#                         |                                |
+#                make_model(DPAtomicModel)                 |
+#                         |                                |
+#                         ----------------------------------
+#                                           |
+#                                         DPModel
+#
+# The order is: DPModel -> make_model(DPAtomicModel) -> DPAtomicModel ->
+# torch.nn.Module -> BaseAtomicModel -> BaseBaseModel
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pt/model/model/model.py:43]:
```suggestion
# torch.nn.Module -> BaseAtomicModel -> BaseModel -> BaseBaseModel
```

**Comment by wanghan-iapcm** [comment]
> We cannot store the plugins of dpmodel and pt in the same class. There must be seperately classes to initialize the plugin separately.

it should be possible to `make_base_model` for different backends, see #3340

File: deepmd/dpmodel/model/base_model.py [hunk 7]
```diff
@@ -6,136 +6,142 @@
...
-        pass
-
-    @abstractmethod
-    def get_type_map(self) -> List[str]:
-        """Get the type map."""
-
-    @abstractmethod
-    def get_rcut(self):
-        """Get the cut-off radius."""
+def make_base_model() -> Type[object]:
```
Comment (njzjz) [inline deepmd/dpmodel/model/base_model.py:18]: Added `make_base_model`.

File: deepmd/utils/plugin.py [hunk 8]
```diff
@@ -93,3 +96,60 @@ class PluginVariant(metaclass=VariantABCMeta):
     """A class to remove `type` from input arguments."""
 
     pass
+
+
+def make_plugin_registry(name: Optional[str] = None) -> Type[object]:
```
Comment (njzjz) [inline deepmd/utils/plugin.py:101]: I also added a `make_plugin_registry` method and will apply it to other classes after this PR.

File: deepmd/dpmodel/model/base_model.py [hunk 9]
```diff
@@ -151,38 +157,6 @@ class BaseModel(BaseBaseModel):
...
-            cls = cls.get_class_by_type(j_get_type(kwargs, cls.__name__))
-        return super().__new__(cls)
-
-    @classmethod
-    def get_class_by_type(cls, model_type: str) -> Type["BaseModel"]:
-        if model_type in BaseModel.__plugins.plugins:
-            return BaseModel.__plugins.plugins[model_type]
-        else:
-            raise RuntimeError("Unknown model type: " + model_type)
+    def get_model_def_script(self) -> str:
```
Comment (wanghan-iapcm) [inline deepmd/dpmodel/model/base_model.py:160]: The method may be moved to make base atomic model, the same designed as the pt backend.

**Review by wanghan-iapcm — APPROVED** [review]

**Pull request merged**